components/fxa-client/ios/FxAClient/FxAccountManager.swift

```diff
@@ -0,0 +1,584 @@
+        setupAuthExceptionsListener()
+    }
+
+    lazy var statePersistenceCallback: FxAStatePersistenceCallback = {
+        FxAStatePersistenceCallback(manager: self)
+    }()
+
+    public func initialize(completionHandler: @escaping (Result<Void, Error>) -> Void) {
+        processEvent(event: .initialize) {
+            DispatchQueue.main.async { completionHandler(Result.success(())) }
+        }
+    }
+
+    public func hasAccount() -> Bool {
+        return state == .authenticatedWithProfile ||
+            state == .authenticatedNoProfile ||
+            state == .authenticationProblem
+    }
+
+    public func accountNeedsReauth() -> Bool {
+        return state == .authenticationProblem
+    }
+
+    public func beginAuthentication(completionHandler: @escaping (Result<URL, Error>) -> Void) {
+        DispatchQueue.global().async {
+            let result = self.updatingLatestAuthState {
+                try self.handleAccountAuthErrors { acct in
+                    try acct.beginOAuthFlowSync(scopes: self.applicationScopes)
+                }
+            }
+            DispatchQueue.main.async { completionHandler(result) }
+        }
+    }
+
+    public func beginPairingAuthentication(
+        pairingUrl: String,
+        completionHandler: @escaping (Result<URL, Error>) -> Void
+    ) {
+        DispatchQueue.global().async {
+            let result = self.updatingLatestAuthState {
+                try self.handleAccountAuthErrors { acct in
+                    try acct.beginPairingFlowSync(pairingUrl: pairingUrl, scopes: self.applicationScopes)
+                }
+            }
+            DispatchQueue.main.async { completionHandler(result) }
+        }
+    }
+
+    private func updatingLatestAuthState(_ beginFlowFn: () throws -> URL) -> Result<URL, Error> {
+        do {
+            let url = try beginFlowFn()
+            let comps = URLComponents(url: url, resolvingAgainstBaseURL: true)
+            latestAuthState = comps!.queryItems!.first(where: { $0.name == "state" })!.value
+            return .success(url)
+        } catch {
+            return .failure(error)
+        }
+    }
+
+    public func finishAuthentication(
+        authData: FxaAuthData,
+        completionHandler: @escaping (Result<Void, Error>) -> Void
+    ) {
+        if latestAuthState == nil {
+            DispatchQueue.main.async { completionHandler(.failure(FirefoxAccountError.noExistingAuthFlow)) }
+        } else if authData.state != latestAuthState {
+            DispatchQueue.main.async { completionHandler(.failure(FirefoxAccountError.wrongAuthFlow)) }
+        } else { /* state == latestAuthState */
+            processEvent(event: .authenticated(authData: authData)) {
+                DispatchQueue.main.async { completionHandler(.success(())) }
+            }
+        }
+    }
+
+    public func getAccessToken(scope: String, completionHandler: @escaping (Result<AccessTokenInfo, Error>) -> Void) {
```

Comment:
It's not obvious to me why some of the methods on this class dispatch via `DispatchQueue.global().async` and others don't. As a consumer of this class, is it important that I know which ones do and do not? If so, is there a convention we could use to help make that clearer?

Reply:
@garvankeeley if you have any insights here?